### PR TITLE
FEATURE: add keyboard navigation for serachable Dropdowns

### LINKED ISSUE #1104: Metadata in Inline Link Editor is not legible on hover
**[<- must-have for 1.0.0](https://github.com/neos/neos-ui/issues/687)**

screenshot_2017-10-05_08-32-41

### LINKED ISSUE #1162: Add Keyboard Navigation for searchable dropdowns
**[<- must-have for 1.0.0](https://github.com/neos/neos-ui/issues/687)**

### PR DESCRIPTION
**Pull request by mstruebing**
This enables keyboard navigation for searchable dropdown editors and also for the link editor

closes #1162 
closes #1104

### CODE REVIEW

File: packages/react-ui-components/src/SelectBox/selectBox.js [hunk 1]
```diff
@@ -158,40 +156,50 @@ export default class SelectBox extends PureComponent {
...
         const listener = e => {
-            // Up and down arrow, enter
-            if ([38, 40, 13].indexOf(e.keyCode) > -1) {
+            const UP_ARROW = 38;
+            const DOWN_ARROW = 40;
+            const ENTER = 13;
+
+            // debouncing with lodash doesn't work as intended
+            // so I had to write my own debouncing function
+            const debounced = () => Date.now() - this.state.lastCall > 50;
```
Comment (dimaip) [inline packages/react-ui-components/src/SelectBox/selectBox.js]: Denouncing via `setState`? But why? Isn't there a way to do it properly?

File: packages/react-ui-components/src/SelectBox/selectBox.js [hunk 2]
```diff
@@ -158,40 +156,50 @@ export default class SelectBox extends PureComponent {
 
     componentDidUpdate() {
         const listener = e => {
```
Comment (dimaip) [inline packages/react-ui-components/src/SelectBox/selectBox.js]: Is it needed to define a listener function on each update? Maybe set it as a class method or move it outside of the class declaration altogether?

File: packages/react-ui-components/src/SelectBox/selectBox.js (hunk 1 above)
Comment (mstruebing) [reply packages/react-ui-components/src/SelectBox/selectBox.js]: https://github.com/neos/neos-ui/pull/1165/commits/cd67d221d7a8f1525f5a2c9912115445b956acf6#diff-46984f3d5a9788f9c39ad32b4dadc224R200 this doesn't work as I wanted it to :/

File: packages/react-ui-components/src/SelectBox/selectBox.js [hunk 3]
```diff
@@ -154,20 +155,65 @@ export default class SelectBox extends PureComponent {
...
-            window.addEventListener('keydown', listener);
+            window.addEventListener('keydown', this.keydownListener);
         } else {
-            window.removeEventListener('keydown', listener);
+            window.removeEventListener('keydown', this.keydownListener);
         }
     }
 
+    keydownListener = debounce((e) => {
+        const UP_ARROW = 38;
```
Comment (dimaip) [inline packages/react-ui-components/src/SelectBox/selectBox.js]: I think you could use import key definitions like this? https://www.npmjs.com/package/react-keydown#specify-multiple-keys-that-should-trigger-the-method

File: packages/react-ui-components/src/SelectBox/selectBox.js [hunk 4]
```diff
@@ -2,10 +2,12 @@ import React, {PureComponent} from 'react';
 import PropTypes from 'prop-types';
 import DropDown from '../DropDown/index';
 import DefaultSelectBoxOption from './defaultSelectBoxOption';
-import keydown from 'react-keydown';
+import keydown, {Keys} from 'react-keydown';
 import mergeClassNames from 'classnames';
+import debounce from 'lodash.debounce';
 
-const KEYS = ['down', 'up', 'enter'];
+const { ENTER, UP, DOWN} = Keys;
```
Comment (dimaip) [inline packages/react-ui-components/src/SelectBox/selectBox.js]: extra space

**Pull request merged**